code/sonalyze/sonarlog/ingest.go

```diff
@@ -0,0 +1,44 @@
+package sonarlog
+
+import (
+	"time"
+
+	"go-utils/config"
+	"go-utils/hostglob"
+	"sonalyze/db"
+)
+
+// This may become more complicated as we implement caching.  In particular, each db.LogFile will
+// cache a []*db.Sample where:
+//
+//  - there is no CpuUtilPct field (this is crucial)
+//  - time-duplicate records have been removed (not crucial but fairly desirable)
+//  - the GPU memory fixups have been applied (not crucial but fairly desirable)
+//
+// This means that some of the logic of createInputStreams will move into the db layer and that the
+// input layer will return not one total list of samples but probably one list per file (a list of
+// lists), and that the job here will be to stich those together and resolve any issues at the
+// joins.  How we do this depends on where we want to put the complexity and how much redundant
+// computation we want to avoid.  It's possible to apply the GPU fixups without first constructing
+// InputStreamSets, for example.
+
+func ReadSampleStreams(
+	c db.SampleCluster,
+	fromDate, toDate time.Time,
+	hostGlobber *hostglob.HostGlobber,
+	cfg *config.ClusterConfig,
+	verbose bool,
+) (
+	streams InputStreamSet,
+	bounds Timebounds,
+	read, dropped int,
+	err error,
+) {
+	samples, dropped, err := c.ReadSamples(fromDate, toDate, hostGlobber, verbose)
+	if err != nil {
+		return
+	}
+	read = len(samples)
+	streams, bounds = createInputStreams(samples, cfg)
```

Comment:
This is not great.  createInputStreams will side-effect the individual samples, overwriting the gpu memory fields.  If the cfg is always the same (this is not a given - and the caches should probably be purged for a cluster if the cluster config is found to change, not currently a problem but a problem if we implement HUP handling) then we will overwrite with the same data, but this is still a data race.  The only way to really fix this is to push that part of the code into the reading code (when new data are read the gpu fixups are applied) or at least into ReadSamples (which could maybe discover if a request was filled from cache, and apply fixups).

This ties into a comment elsewhere about enforcing the presence of a config file.

Reply:
I have "fixed" this with a comment but will investigate further.